feat: a lowering pass to re-compose ops into aten.linear

zewenli98 opened the pull request:
# Description

A lowering pass to re-compose ops into `aten.linear`.

A discussion about why we need this lowering pass can be found here: #2296 

Fixes #2398 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

gs-olive commented on py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py:
```diff
@@ -0,0 +1,67 @@
+):
+    """Constructs the original and replacement functions for linear"""
+
+    # Empty boilerplate function taking in three Tensors and returning one
+    def boilerplate(
+        input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor
+    ) -> torch.Tensor:
+        ...
+
+    # Trace boilerplate function and extract placeholder and output nodes
+    orig = torch.fx.symbolic_trace(boilerplate)
+    input, weight, bias = get_tensor_placeholders(orig)
+    output = [node for node in orig.graph.nodes if node.op == "output"][0]
+
+    with orig.graph.inserting_before(output):
+        W_T = orig.graph.call_function(
+            torch.ops.aten.permute.default,
+            args=(weight, [1, 0]),
+        )
+        out = orig.graph.call_function(
+            torch.ops.aten.addmm.default,
+            args=(bias, input, W_T),
+        )
+
+    # Assign the output of the graph to be the single getitem output
+    output.args = (out,)
+
+    orig.graph.lint()
+    orig.recompile()
```
The `linear` operator may not necessarily have the same restrictions as the attention operator did, so it may not be necessary to build the fx graph from scratch. For instance, this whole block could instead be:
```python
    def orig(
        input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor
    ) -> torch.Tensor:
        W_T = torch.ops.aten.permute.default(weight, [1, 0])
        out = torch.ops.aten.addmm.default(bias, input, W_T)
        return out
```
The above would need verification to validate functionality.

zewenli98 replied:
Thanks for the details. I'm thinking why we need these restrictions for `attention` but not for `linear`. Is there anything different in `attention`? Can we just define the `orig` like:
```python
def orig(
        q: torch.Tensor, k: torch.Tensor, v: torch.Tensor
    ) -> torch.Tensor:
        outputs = torch.ops.aten._scaled_dot_product_efficient_attention.default(q, k, v, None, False)
        out = operator.getitem(outputs, 0)
        return out
```

zewenli98 replied:
I just tried to use the `orig` function above, and it also works for `attention`.

gs-olive replied:
Ok that's great - the above is definitely cleaner and the preferred usage. I think I had overlooked it or was working on a Torch nightly where that wasn't the case. If you could add the change to the attention pass as well for this PR, that would be much appreciated.

gs-olive commented on tests/py/dynamo/lowering/test_aten_lowering_passes.py:
```diff
@@ -267,5 +267,110 @@ def forward(self, q, k, v):
+
+        fx_graph = torch.fx.symbolic_trace(Linear())
+        expected_ops = {torch.ops.aten.linear.default}
+        unexpected_ops = {}
```
Unexpected ops include `addmm` and `permute`

gs-olive commented on py/torch_tensorrt/dynamo/lowering/_decompositions.py:
```diff
@@ -105,22 +105,6 @@ def alias_replacement(x: torch.Tensor) -> torch.Tensor:
     return x
 
 
-@register_torch_trt_decomposition(
-    torch.ops.aten.addmm, registry=TORCH_TRT_DECOMPOSITIONS
-)
-def addmm_replacement(
-    input_: torch.Tensor,
-    mat1: torch.Tensor,
-    mat2: torch.Tensor,
-    *,
-    beta: int = 1,
-    alpha: int = 1,
-) -> torch.Tensor:
-    return torch.add(
-        torch.mul(input_, beta), torch.mul(torch.matmul(mat1, mat2), alpha)
-    )
```
This PR also requires a converter for `aten.addmm`, since we are no longer decomposing it, so it would be a decrease in coverage otherwise

gs-olive replied:
Additionally, you can remove the `addmm` lowering pass tests, since the functionality is no longer present

gs-olive reviewed (changes requested):
Added comments detailing suggested improvements. Overall the design looks good!

zewenli98 commented:
@gs-olive Thanks for your review and constructive comments. I have addressed all the issues above. Please take a look when you get the chance. Thanks a lot!

zewenli98 commented:
@gs-olive Can we merge into the main branch if everything looks good?

gs-olive commented:
@zewenli98 - this one looks good, but would like to wait until later this week to merge this particular change

gs-olive reviewed (approved):
Looks good to me!
